File: generation/dnp3/src/main/scala/com/automatak/render/dnp3/objects/groups/Group41.scala
/**
 * Copyright 2013-2020 Automatak, LLC
 *
 * Licensed to Green Energy Corp (www.greenenergycorp.com) and Automatak
 * LLC (www.automatak.com) under one or more contributor license agreements.
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership. Green Energy Corp and Automatak LLC license
 * this file to you under the Apache License, Version 2.0 (the "License"); you
 * may not use this file except in compliance with the License. You may obtain
 * a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.automatak.render.dnp3.objects.groups

import com.automatak.render.dnp3.objects._

import FixedSizeField._
import com.automatak.render.dnp3.objects.VariationNames._
import com.automatak.render.dnp3.objects.generators.{ConversionToAnalogOutputDouble64, ConversionToAnalogOutputFloat32, ConversionToAnalogOutputInt32, ConversionToAnalogOutputInt16}


object Group41 extends ObjectGroup {
  def objects = List(Group41Var0, Group41Var1, Group41Var2, Group41Var3, Group41Var4)
  def group: Byte = 41
  def desc: String = "Analog Output"
  def isEventGroup: Boolean = false
}

object Group41Var0 extends AnyVariation(Group41, 0)
object Group41Var1 extends FixedSize(Group41, 1, bit32WithFlag)(value32, commandStatus) with ConversionToAnalogOutputInt32
object Group41Var2 extends FixedSize(Group41, 2, bit16WithFlag)(value16, commandStatus) with ConversionToAnalogOutputInt16
object Group41Var3 extends FixedSize(Group41, 3, singlePrecision)(float32, commandStatus) with ConversionToAnalogOutputFloat32
object Group41Var4 extends FixedSize(Group41, 4, doublePrecision)(float64, commandStatus) with ConversionToAnalogOutputDouble64

File: cpp/opendnp3/src/opendnp3/objects/Group41.cpp
//
//  _   _         ______    _ _ _   _             _ _ _
// | \ | |       |  ____|  | (_) | (_)           | | | |
// |  \| | ___   | |__   __| |_| |_ _ _ __   __ _| | | |
// | . ` |/ _ \  |  __| / _` | | __| | '_ \ / _` | | | |
// | |\  | (_) | | |___| (_| | | |_| | | | | (_| |_|_|_|
// |_| \_|\___/  |______\__,_|_|\__|_|_| |_|\__, (_|_|_)
//                                           __/ |
//                                          |___/
// Copyright 2013 Automatak LLC
// 
// Automatak LLC (www.automatak.com) licenses this file
// to you under the the Apache License Version 2.0 (the "License"):
// 
// http://www.apache.org/licenses/LICENSE-2.0.html
//

#include "Group41.h"

#include "../PackingUnpacking.h"

namespace opendnp3 {

Group41Var1 Group41Var1::Read(openpal::ReadOnlyBuffer& buffer)
{
  Group41Var1 obj;
  obj.value = Int32LE::Read(buffer);
  buffer.Advance(4);
  obj.status = CommandStatusFromType(*buffer);
  buffer.Advance(1);
  return obj;
}

Group41Var2 Group41Var2::Read(openpal::ReadOnlyBuffer& buffer)
{
  Group41Var2 obj;
  obj.value = Int16LE::Read(buffer);
  buffer.Advance(2);
  obj.status = CommandStatusFromType(*buffer);
  buffer.Advance(1);
  return obj;
}

Group41Var3 Group41Var3::Read(openpal::ReadOnlyBuffer& buffer)
{
  Group41Var3 obj;
  obj.value = SingleFloat::Read(buffer);
  buffer.Advance(4);
  obj.status = CommandStatusFromType(*buffer);
  buffer.Advance(1);
  return obj;
}

Group41Var4 Group41Var4::Read(openpal::ReadOnlyBuffer& buffer)
{
  Group41Var4 obj;
  obj.value = SingleFloat::Read(buffer);
  buffer.Advance(4);
  obj.status = CommandStatusFromType(*buffer);
  buffer.Advance(1);
  return obj;
}


}
